🐛 Fix build contamination in lgpio cross-compilation script

## Problem Statement

The lgpio cross-compilation script (`scripts/setup-lgpio-cross.sh`) has a latent build contamination bug where 32-bit and 64-bit builds share the same build directory. This causes make to skip recompilation when switching architectures, leading to linking errors.

**Current Status:**
```bash
# scripts/setup-lgpio-cross.sh:16
BUILD_DIR="${BUILD_DIR:-/tmp/lg-build}"  # ← Same for both architectures!
```

**Impact:**
- ⚠️ 64-bit build inherits 32-bit object files
- ⚠️ Make skips recompilation (timestamps indicate files are "up to date")
- ❌ Linker fails with "file in wrong format" error
- ⚠️ Intermittent CI failures possible

## Root Cause Analysis

### Make's Timestamp-Based Dependencies

Make's dependency tracking is based on **file timestamps**, not compiler detection:

```makefile
# lgpio/Makefile
CROSS_PREFIX =
CC = $(CROSS_PREFIX)gcc

lgpio.o: lgpio.c lgpio.h
    $(CC) $(CFLAGS) -c -o lgpio.o lgpio.c
```

**Make's logic:**
1. Check: Is `lgpio.o` newer than `lgpio.c`?
2. If YES → Skip compilation
3. **Problem:** Make doesn't know that changing CROSS_PREFIX requires rebuild

### Reproduction

**Confirmed in testing:**
```bash
# First build (32-bit):
$ cd /tmp/lg-build/lg
$ make CROSS_PREFIX=arm-linux-gnueabihf- lib
arm-linux-gnueabihf-gcc -c -o lgCtx.o lgCtx.c  # ✓ Compiles 32-bit

# Second build (64-bit) - SAME directory:
$ make CROSS_PREFIX=aarch64-linux-gnu- lib
make: Nothing to be done for target 'lib'  # ❌ Skips compilation!

# Verify the problem:
$ file lgCtx.o
lgCtx.o: ELF 32-bit LSB relocatable, ARM, EABI5 version 1

# Try to link with 64-bit compiler:
$ aarch64-linux-gnu-gcc -shared -o test.so lgCtx.o
error adding symbols: file in wrong format
ld returned 1 exit status  # ❌ LINKING FAILS
```

### Why It Happens

**Current script logic (lines 56-66):**
```bash
if [ -d "lg/.git" ]; then
    echo "Updating existing lg repository..."
    cd lg
    git pull
    echo "Cleaning previous build artifacts..."
    make clean  # ← Only if repo already exists
else
    echo "Cloning lg repository..."
    git clone https://github.com/joan2937/lg.git
    cd lg  # ← No cleaning on fresh clone
fi

# Build immediately - may see old .o files from different architecture
make CROSS_PREFIX="$CROSS_PREFIX" lib
```

**Problem:** 
- `make clean` runs AFTER the first architecture builds
- On the second architecture build, make sees .o files with newer timestamps than .c files
- Make skips compilation, keeping the wrong architecture's object files

## Current Workflow Behavior

The CI workflow runs both architectures sequentially **in the same job**:

```yaml
# .github/workflows/examples.yml
- name: Build lgpio for ARM cross-compilation (32-bit)
  run: ./scripts/setup-lgpio-cross.sh

- name: Build lgpio for ARM cross-compilation (64-bit)
  run: |
    CROSS_PREFIX=aarch64-linux-gnu- \
    INSTALL_DIR=$HOME/.local/aarch64-linux-gnu \
    ./scripts/setup-lgpio-cross.sh
```

**Result:** Both builds use `/tmp/lg-build` → contamination risk!

## Proposed Solution

**Option A: Separate Build Directories** (Recommended)

Use architecture-specific build directories to ensure complete isolation:

```bash
# scripts/setup-lgpio-cross.sh:16-25
# Detect architecture from cross-compiler prefix and set arch-specific build dir
if [[ "$CROSS_PREFIX" == "aarch64-"* ]]; then
    ARCH_NAME="ARM 64-bit (AArch64)"
    BUILD_DIR="${BUILD_DIR:-/tmp/lg-build-aarch64}"  # ← Separate for 64-bit
else
    ARCH_NAME="ARM 32-bit (ARMv7)"
    BUILD_DIR="${BUILD_DIR:-/tmp/lg-build-armhf}"   # ← Separate for 32-bit
fi
```

**Benefits:**
- ✅ Complete isolation between architectures
- ✅ No risk of contamination
- ✅ Matches professional build system patterns
- ✅ Consistent with recommended fix for libgpiod (#86)

**Option B: Force Clean Rebuild**

Always start with a clean build directory:

```bash
# Nuclear option - remove and recreate build directory
echo "Cleaning build directory..."
rm -rf "$BUILD_DIR"
mkdir -p "$BUILD_DIR"
cd "$BUILD_DIR"

echo "Cloning lg repository..."
git clone https://github.com/joan2937/lg.git
cd lg
```

**Benefits:**
- ✅ Guarantees clean build every time
- ✅ Simple to implement

**Drawbacks:**
- ❌ Slower (re-clones every time)
- ❌ More network bandwidth

**Option C: Aggressive Git Cleaning**

Use `git clean` to remove all untracked files:

```bash
if [ -d "lg/.git" ]; then
    echo "Updating existing lg repository..."
    cd lg
    git clean -fdx  # Remove ALL untracked files and build artifacts
    git reset --hard  # Reset modifications
    git pull
else
    echo "Cloning lg repository..."
    git clone https://github.com/joan2937/lg.git
    cd lg
fi
```

**Benefits:**
- ✅ Ensures clean state
- ✅ Faster than re-cloning

**Drawbacks:**
- ❌ Still shares directory (potential for race conditions if builds run in parallel)

## Recommended Implementation

**Option A** (separate directories) is recommended because:
1. Provides complete isolation
2. No risk of contamination
3. Matches standard build system practices
4. Consistent with libgpiod fix (#86)
5. Enables parallel builds in the future

## Implementation Tasks

- [ ] Update `scripts/setup-lgpio-cross.sh`
  - [ ] Add architecture detection to determine build directory
  - [ ] Set `BUILD_DIR` based on architecture (armhf vs aarch64)
  - [ ] Update documentation comments

- [ ] Testing
  - [ ] Verify 32-bit build creates correct directory
  - [ ] Verify 64-bit build creates separate directory
  - [ ] Confirm no cross-contamination between builds
  - [ ] Verify lgpio examples build successfully
  - [ ] Test both sequential and parallel builds

- [ ] Documentation
  - [ ] Update script comments explaining the architecture-specific directories
  - [ ] Document in commit message why this change was necessary

## Acceptance Criteria

✅ **Isolation:**
- 32-bit builds use `/tmp/lg-build-armhf`
- 64-bit builds use `/tmp/lg-build-aarch64`
- No shared build artifacts between architectures

✅ **Correctness:**
- Both architectures build successfully
- Object files are correct architecture (`file *.o` shows correct ELF type)
- Libraries link correctly
- No "file in wrong format" errors

✅ **CI/CD:**
- Workflow builds both architectures without errors
- `examples/lgpio-blink` compiles for both armhf and aarch64
- No contamination-related failures

## Related Issues

- #85 - Build libgpiod from source on Ubuntu (similar contamination issue)
- #86 - PR implementing libgpiod fix (should include same pattern)

This issue was discovered during code review of PR #86, where the same contamination bug was identified in the libgpiod script. Both scripts should be fixed with the same pattern for consistency.

## Priority

**P2 - Medium Priority**

**Rationale:**
- Latent bug (not currently causing CI failures)
- But could cause intermittent failures in specific scenarios
- Should be fixed for consistency with libgpiod fix
- Prevents future debugging headaches
- Simple fix with clear solution

## Testing Evidence

**Test performed on 2025-11-15:**
```bash
$ cd /tmp && git clone https://github.com/joan2937/lg.git lg-test
$ cd lg-test
$ make CROSS_PREFIX=arm-linux-gnueabihf- lib
# Compiles successfully

$ make CROSS_PREFIX=aarch64-linux-gnu- lib
make: Nothing to be done for target 'lib'  # ❌ BUG CONFIRMED

$ file lgCtx.o
lgCtx.o: ELF 32-bit LSB relocatable, ARM, EABI5 version 1  # Wrong arch!

$ aarch64-linux-gnu-gcc -shared -o test.so lgCtx.o
error adding symbols: file in wrong format  # ❌ Linking fails
```

This confirms the bug exists and will cause linking failures when switching architectures without proper directory isolation.

--- a/scripts/setup-lgpio-cross.sh
+++ b/scripts/setup-lgpio-cross.sh
@@ -13,15 +13,18 @@ echo ""
 # Resolve script and repo directories BEFORE any cd commands
 SCRIPT_DIR="$(cd "$(dirname "${BASH_SOURCE[0]}")" && pwd)"
 REPO_ROOT="$(dirname "$SCRIPT_DIR")"
-BUILD_DIR="${BUILD_DIR:-/tmp/lg-build}"
 CROSS_PREFIX="${CROSS_PREFIX:-arm-linux-gnueabihf-}"
 INSTALL_DIR="${INSTALL_DIR:-$HOME/.local/arm-linux-gnueabihf}"
 
-# Detect architecture from cross-compiler prefix
+# Detect architecture from cross-compiler prefix and set arch-specific build dir
+# This prevents build contamination when switching between 32-bit and 64-bit builds
+# (make's timestamp-based dependency tracking doesn't detect compiler changes)
 if [[ "$CROSS_PREFIX" == "aarch64-"* ]]; then
     ARCH_NAME="ARM 64-bit (AArch64)"
+    BUILD_DIR="${BUILD_DIR:-/tmp/lg-build-aarch64}"
 else
     ARCH_NAME="ARM 32-bit (ARMv7)"
+    BUILD_DIR="${BUILD_DIR:-/tmp/lg-build-armhf}"
 fi
 
 # Check if cross-compiler is installed

